src/solvers/smt2/smt2_conv.cpp

```diff
@@ -4068,38 +4051,98 @@ void smt2_convt::set_to(const exprt &expr, bool value)
 
         convert_type(equal_expr.lhs().type());
         out << " ";
-        convert_expr(equal_expr.rhs());
+        convert_expr(prepared_rhs);
 
         out << ")" << "\n";
         return; // done
       }
     }
   }
 
-  find_symbols(expr);
+  exprt prepared_expr = prepare_expr(expr);
 
-  #if 0
+#if 0
   out << "; CONV: "
       << format(expr) << "\n";
-  #endif
+#endif
 
   out << "; set_to " << (value?"true":"false") << "\n"
       << "(assert ";
 
   if(!value)
   {
     out << "(not ";
-    convert_expr(expr);
+    convert_expr(prepared_expr);
     out << ")";
   }
   else
-    convert_expr(expr);
+    convert_expr(prepared_expr);
 
   out << ")" << "\n"; // assert
 
   return;
 }
 
+/// Lower byte_update and byte_extract operations within \p expr. Return an
+/// equivalent expression that doesn't use byte operators.
+/// Note this replaces operators post-order (compare \ref lower_byte_operators,
+/// which uses a pre-order walk, replacing in child expressions before the
+/// parent). Pre-order replacement currently fails regression tests: see
+/// https://github.com/diffblue/cbmc/issues/4380
+exprt smt2_convt::lower_byte_operators_rec(const exprt &expr)
+{
+  exprt lowered_expr = expr;
+
+  for(auto it = lowered_expr.depth_begin(), itend = lowered_expr.depth_end();
+      it != itend;
+      /* No ++it */)
+  {
+    if(
+      it->id() == ID_byte_extract_little_endian ||
+      it->id() == ID_byte_extract_big_endian)
+    {
+      // Note we must recurse to get a fresh iterator because depth_iteratort
+      // currently crashes if we try to recurse into an expression we've
+      // altered.
+      it.mutate() = lower_byte_operators_rec(
+        lower_byte_extract(to_byte_extract_expr(*it), ns));
+      it.next_sibling_or_parent();
+    }
+    else if(
+      it->id() == ID_byte_update_little_endian ||
+      it->id() == ID_byte_update_big_endian)
+    {
+      // Note we must recurse to get a fresh iterator because depth_iteratort
+      // currently crashes if we try to recurse into an expression we've
+      // altered.
+      it.mutate() = lower_byte_operators_rec(
+        lower_byte_update(to_byte_update_expr(*it), ns));
+      it.next_sibling_or_parent();
+    }
+    else
+    {
+      ++it;
+    }
+  }
+
+  return lowered_expr;
+}
+
+exprt smt2_convt::prepare_expr(const exprt &expr)
```

Comment:
Please document this function. Also a better name than prepare_expr would be helpful. 
I am a bit surprised this and lower_byte_operators_rec are not const, and if they could be made static this would be better, in particular you could avoid exposing auxiliary functions in the header.

Reply:
Documented and renamed. The whole class uses member functions to pass the namespace around -- I won't get into changing that in this PR.